wit/streams.wit

```diff
@@ -259,4 +259,40 @@ interface streams {
             len: u64,
         ) -> result<u64, stream-error>;
     }
+
+    /// Continuously read all data from the input stream and write it to the
+    /// output stream.
+    ///
+    /// Forwarding is done on a background task. The tasks runs until either
+    /// stream closes or fails. After that, the streams are dropped and the
+    /// returned future is resolved. Dropping the future before it has been
+    /// resolved aborts the background task and drops the streams.
+    ///
+    /// The future resolves with the reason that caused the forwarding to terminate.
+    /// In case one of the streams ended normally, this will be `stream-error::closed`.
+    ///
+    /// This function requires exclusive access to the provided streams,
+    /// yet it does not change the ownership structure of them. I.e. they remain
+    /// children of their parent resources. If the parents place any lifetimes
+    /// restrictions on their children, those continue to apply.
+    forward: func(src: input-stream, dst: output-stream) -> future-forward-result;
```

Comment:
These arguments will pass ownership of both the input-stream and output-stream. I believe sending an owned input-stream is correct here. In some use cases, we need a way to permit further use of the output-stream after forwarding has either finished or failed. We can do that by either making this parameter a borrow of the output-stream, or we get an owned output-stream in the resolution of future-forward-result.get

Reply:
Changing `dst: output-stream` from an owned handle to a borrowed handle will introduce multithreaded access to the stream. I can't see a way why one would want this as there's no way coordinate between the writes from different tasks/threads. From end-user and implementation complexity perspective I'd go for: consume the stream and hand it back later. Or even simpler: document & enforce that while the forward is in progress any attempt to access the borrowed streams will trap.

That being said, the way I designed it, the host is guaranteed that no one else can touch the streams ever again. In my mind this is a "feature", as the host can then permanently fuse them together. This may enable additional optimizations not possible otherwise.

An example that triggered me to file this PR in the first place is: kTLS. Piping a TLS stream into a Socket stream could "automagically" offload TLS to the kernel (or even the NIC). Requiring the output-stream to remain usable after the input-stream has ended would prevent this optimization. To be clear, I haven't actually implemented this, so assign as much value to this use-case as you want.

Anyway, I can see merit in both designs. Maybe they can coexist?

Reply:
I hadn't thought about that case, but I have heard of some use cases for wanting to use the destination stream after  the source stream is consumed. So, I guess they can both coexist. Are you up for introducing both variants in this PR?